turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -1,33 +1,102 @@
+    graph::{AdjacencyMap, GraphTraversal, Visit, VisitControlFlow},
     trace::{TraceRawVcs, TraceRawVcsContext},
-    NonLocalValue, ReadRef, ResolvedVc, TryJoinIterExt, ValueToString, Vc,
+    FxIndexMap, NonLocalValue, ReadRef, ResolvedVc, TryJoinIterExt, ValueToString, Vc,
 };
 
 use crate::{
     chunk::ChunkingType,
-    issue::{Issue, IssueExt},
+    issue::Issue,
     module::{Module, Modules},
     reference::primary_chunkable_referenced_modules,
 };
 
+#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq, Serialize, Deserialize, TraceRawVcs)]
+pub struct GraphNodeIndex {
+    #[turbo_tasks(trace_ignore)]
+    graph_idx: usize,
+    #[turbo_tasks(trace_ignore)]
+    node_idx: NodeIndex,
+}
+unsafe impl NonLocalValue for GraphNodeIndex {}
+
 #[turbo_tasks::value(transparent)]
 #[derive(Clone, Debug)]
-pub struct ModuleSet(pub HashSet<ResolvedVc<Box<dyn Module>>>);
+pub struct VisitedModules(pub FxIndexMap<ResolvedVc<Box<dyn Module>>, GraphNodeIndex>);
+
+#[turbo_tasks::value_impl]
+impl VisitedModules {
+    #[turbo_tasks::function]
+    pub async fn empty() -> Vc<Self> {
+        Vc::cell(FxIndexMap::default())
+    }
+
+    #[turbo_tasks::function]
+    pub async fn from_graph(graph: Vc<SingleModuleGraph>) -> Result<Vc<Self>> {
+        Ok(Vc::cell(
+            graph
+                .await?
+                .enumerate_nodes()
+                .flat_map(|(node_idx, module)| match module {
+                    SingleModuleGraphNode::Module(SingleModuleGraphModuleNode {
+                        module, ..
+                    }) => Some((
+                        *module,
+                        GraphNodeIndex {
+                            graph_idx: 0,
+                            node_idx,
+                        },
+                    )),
+                    SingleModuleGraphNode::VisitedModule { .. } => None,
+                })
+                .collect(),
+        ))
+    }
+
+    #[turbo_tasks::function]
+    pub async fn concatenate(&self, graph: Vc<SingleModuleGraph>) -> Result<Vc<Self>> {
+        let graph = graph.await?;
+        let iter = self.0.iter().map(|(module, idx)| (*module, *idx)).chain(
+            graph
+                .enumerate_nodes()
+                .flat_map(|(node_idx, module)| match module {
+                    SingleModuleGraphNode::Module(SingleModuleGraphModuleNode {
+                        module, ..
+                    }) => Some((
+                        *module,
+                        GraphNodeIndex {
+                            graph_idx: 0,
+                            node_idx,
+                        },
+                    )),
+                    SingleModuleGraphNode::VisitedModule { .. } => None,
+                }),
+        );
+
+        // `map.extend()` but don't overwrite keys (give higher priority to the earlier maps)
+        let mut map = FxIndexMap::default();
+        map.reserve(iter.size_hint().0);
```

Comment:
I don't think `flat_map` will give you a useful size_hint(), so it's basically only the self.0.len()